internal/ethereum/event_actions.go

```diff
@@ -0,0 +1,173 @@
+		id:      es.id,
+		ctx:     req.StreamContext,
+		updates: req.BlockListener,
+	})
+
+	return &ffcapi.EventStreamStartResponse{}, "", nil
+}
+
+func (c *ethConnector) EventStreamStopped(ctx context.Context, req *ffcapi.EventStreamStoppedRequest) (*ffcapi.EventStreamStoppedResponse, ffcapi.ErrorReason, error) {
+	c.mux.Lock()
+	es := c.eventStreams[*req.ID]
+	c.mux.Unlock()
+	if es != nil {
+		select {
+		case <-es.ctx.Done():
+			// This is good, it is stopped
+		default:
+			return nil, ffcapi.ErrorReason(""), i18n.NewError(ctx, msgs.MsgStreamNotStopped, req.ID)
+		}
+	}
+	c.mux.Lock()
+	delete(c.eventStreams, *req.ID)
+	listeners := make([]*listener, 0)
+	for _, l := range es.listeners {
+		listeners = append(listeners, l)
+	}
+	c.mux.Unlock()
+	// Wait for stream loop to complete
+	<-es.streamLoopDone
+	// Wait for any listener catchup loops
+	for _, l := range listeners {
+		if l.catchupLoopDone != nil {
+			<-l.catchupLoopDone
+		}
+	}
+	return &ffcapi.EventStreamStoppedResponse{}, "", nil
+}
+
+func (c *ethConnector) EventListenerVerifyOptions(ctx context.Context, req *ffcapi.EventListenerVerifyOptionsRequest) (*ffcapi.EventListenerVerifyOptionsResponse, ffcapi.ErrorReason, error) {
+
+	signature, _, err := parseEventFilters(ctx, req.Filters)
+	if err != nil {
+		return nil, "", err
+	}
+
+	options, err := parseListenerOptions(ctx, req.Options)
+	if err != nil {
+		return nil, "", err
+	}
+
+	ob, _ := json.Marshal(&options)
+	return &ffcapi.EventListenerVerifyOptionsResponse{
+		ResolvedSignature: signature,
+		ResolvedOptions:   fftypes.JSONAny(ob),
+	}, "", nil
+
+}
+
+func (c *ethConnector) EventListenerAdd(ctx context.Context, req *ffcapi.EventListenerAddRequest) (*ffcapi.EventListenerAddResponse, ffcapi.ErrorReason, error) {
+	c.mux.Lock()
+	es := c.eventStreams[*req.StreamID]
+	c.mux.Unlock()
+	if es == nil {
+		return nil, ffcapi.ErrorReasonNotFound, i18n.NewError(ctx, msgs.MsgStreamNotStarted, req.StreamID)
+	}
+	l, err := es.addEventListener(ctx, req)
+	if err != nil {
+		return nil, ffcapi.ErrorReason(""), err
+	}
+	// We start this listener straight away
+	es.startEventListener(l)
+	return &ffcapi.EventListenerAddResponse{}, ffcapi.ErrorReason(""), nil
+}
+
+func (c *ethConnector) EventListenerRemove(ctx context.Context, req *ffcapi.EventListenerRemoveRequest) (*ffcapi.EventListenerRemoveResponse, ffcapi.ErrorReason, error) {
+	c.mux.Lock()
+	es := c.eventStreams[*req.StreamID]
+	c.mux.Unlock()
+	if es == nil {
+		return nil, ffcapi.ErrorReasonNotFound, i18n.NewError(ctx, msgs.MsgStreamNotStarted, req.StreamID)
+	}
+	es.removeEventListener(req.ListenerID)
+	return &ffcapi.EventListenerRemoveResponse{}, ffcapi.ErrorReason(""), nil
+}
+
+func (c *ethConnector) EventStreamNewCheckpointStruct() ffcapi.EventListenerCheckpoint {
+	return &listenerCheckpoint{}
+}
+
+func (c *ethConnector) EventListenerHWM(ctx context.Context, req *ffcapi.EventListenerHWMRequest) (*ffcapi.EventListenerHWMResponse, ffcapi.ErrorReason, error) {
```

Comment:
I think I've figured out that `HWM` stands for High Water Mark. It might be worth a comment somewhere in the code (this may not be the right spot) to describe what that does.